src/main/java/it/pagopa/ecommerce/commons/utils/ConfidentialDataManager.java

```diff
@@ -74,114 +88,155 @@ public interface ConfidentialData {
-     * @throws InvalidAlgorithmParameterException See {@link javax.crypto.Cipher#doFinal(byte[])}
-     * @throws NoSuchPaddingException See {@link javax.crypto.Cipher#doFinal(byte[])}
-     * @throws IllegalBlockSizeException See {@link javax.crypto.Cipher#doFinal(byte[])}
-     * @throws NoSuchAlgorithmException See {@link javax.crypto.Cipher#doFinal(byte[])}
-     * @throws BadPaddingException See {@link javax.crypto.Cipher#doFinal(byte[])}
-     * @throws InvalidKeyException See {@link javax.crypto.Cipher#doFinal(byte[])}
      */
     @Nonnull
-    public <T extends ConfidentialData> String decrypt(Confidential<T> data) throws NoSuchPaddingException, NoSuchAlgorithmException,
-            InvalidAlgorithmParameterException, InvalidKeyException,
-            BadPaddingException, IllegalBlockSizeException {
+    public <T extends ConfidentialData> Mono<String> decrypt(Confidential<T> data) {
 
         return switch (data.confidentialMetadata()) {
-            case AESMetadata aesMetadata -> this.aesCipher.decrypt(aesMetadata, data.opaqueData());
+            case AESMetadata aesMetadata -> {
+                try {
+                    yield Mono.just(this.aesCipher.decrypt(aesMetadata, data.opaqueData()));
+                } catch (NoSuchPaddingException | NoSuchAlgorithmException | InvalidAlgorithmParameterException |
+                         InvalidKeyException | BadPaddingException | IllegalBlockSizeException e) {
+                    throw new ConfidentialDataException(e);
+                }
+            }
+            case PersonalDataVaultMetadata ignored -> this.personalDataVaultClient.findPiiUsingGET(data.opaqueData()).map(PiiResourceDto::getPii);
             default -> throw new IllegalArgumentException("Unsupported cipher metadata!");
         };
     }
 
     @Nonnull
-    private String encryptData(@Nonnull ConfidentialMetadata metadata, @Nonnull String data) throws IllegalBlockSizeException, BadPaddingException, NoSuchPaddingException, NoSuchAlgorithmException, InvalidKeySpecException, InvalidAlgorithmParameterException, InvalidKeyException {
+    private Mono<String> encryptData(@Nonnull ConfidentialMetadata metadata, @Nonnull String data) {
         return switch (metadata) {
-            case AESMetadata aesMetadata -> this.aesCipher.encrypt(aesMetadata, data);
+            case AESMetadata aesMetadata -> {
+                try {
+                    yield Mono.just(this.aesCipher.encrypt(aesMetadata, data));
+                } catch (IllegalBlockSizeException | BadPaddingException | NoSuchPaddingException |
+                         NoSuchAlgorithmException | InvalidAlgorithmParameterException | InvalidKeyException e) {
+                    throw new ConfidentialDataException(e);
+                }
+            }
+            case PersonalDataVaultMetadata ignored -> this.personalDataVaultClient.saveUsingPUT(new PiiResourceDto().pii(data)).map(TokenResourceDto::getToken).map(UUID::toString);
```

Comment:
Catch here the `WebClientResponseException` that could be thrown by the underling client and wrap it to ConfidentialDataException

Reply:
Fixed in [4453ed2](https://github.com/pagopa/pagopa-ecommerce-commons/pull/57/commits/4453ed21820b9d896945f4d9cd4df8bdea37f49b)